src/ws/cockpitauth.c

```diff
@@ -1436,12 +1437,34 @@ cockpit_auth_login_async (CockpitAuth *self,
     }
 
   application = cockpit_auth_parse_application (path, NULL);
-  authorization = cockpit_auth_steal_authorization (headers, connection, &type, &conversation);
 
-  if (!application || !authorization)
+  /* If the client sends a TLS certificate to cockpit-tls, treat this as a
+   * definitive login type, and don't just silently fall back to other types */
+  if (https_instance_has_certificate_file (NULL, 0) != -1)
+    {
+      g_debug ("TLS connection has peer certificate, using tls-cert auth type");
+      type = g_strdup ("tls-cert");
```

Comment:
Should we version this string? Or come up with a completely new one once lis does some TLS handshake foo splitting (we can dream)?

Reply:
For now all the bits that use it are in the same package (cockpit-ws.rpm), and this doesn't work in cockpit/ws container. So we retain the option to change how this works (and I'm sure we'll change it one or three more times). If you feel better with a "tls-cert1" we can do this, but this shines through to cockpit.conf authentication modes, doesn't it? I'd hate to break existing config files some day.